docs/deployment/uwsgi_cookiecutter_nginx.rst

```diff
@@ -0,0 +1,233 @@
+.. _uwsgi_tutorial:
+
+Running a Cookiecutter :app:`Pyramid` Application under :app:`uWSGI` and :app:`Nginx`
+=====================================================================================
+
+``uWSGI`` is a software application for building hosting services.
+It is named after the Web Server Gateway Interface (the WSGI specification
+to which many Python web frameworks conform).
+
+This guide will outline broad steps that can be used to get a Cookiecutter
+:app:`Pyramid` application running under Nginx via ``uWSGI``.  This particular
+tutorial was developed under Ubuntu 18.04, but the instructions should be
+largely the same for all systems, delta specific path information for commands
+and files.
+
+
+#.  Install prerequisites
+
+    .. code-block:: bash
+
+       $ sudo apt install -y uwsgi-core uwsgi-plugin-python3 python3-cookiecutter \
+                             python3-pip python3-venv nginx
+       $ python3 -m cookiecutter gh:Pylons/pyramid-cookiecutter-starter --checkout master
+
+#.  Create a :app:`Pyramid` application. For this tutorial we'll use the
+    ``starter`` :term:`cookiecutter`. See :ref:`project_narr` for more
+    in-depth information about creating a new project.
+
+    .. code-block:: bash
+
+       $ cd
+       $ cookiecutter gh:Pylons/pyramid-cookiecutter-starter --checkout master
+
+    If prompted for the first item, accept the default ``yes`` by hitting return.
+
+    .. code-block:: text
+
+        You've cloned ~/.cookiecutters/pyramid-cookiecutter-starter before.
+        Is it okay to delete and re-clone it? [yes]: yes
+        project_name [Pyramid Scaffold]: myproject
+        repo_name [myproject]: myproject
+        Select template_language:
+        1 - jinja2
+        2 - chameleon
+        3 - mako
+        Choose from 1, 2, 3 [1]: 1
+
+#.  Create a :term:`virtual environment` which we'll use to install our
+    application.
+
+    .. code-block:: bash
+
+       $ cd myproject
+       $ python3 -m venv env
+
+#.  Install your :app:`Pyramid` application and its dependencies.
+
+    .. code-block:: bash
+
+       $ env/bin/pip install -e ".[testing]"
+
+#.  Within the project directory (``~/myproject``), create a script
+    named ``wsgi.py``.  Give it these contents:
+
+    .. code-block:: python
+
+        # Adapted from PServeCommand.run in site-packages/pyramid/scripts/pserve.py
+        from pyramid.scripts.common import get_config_loader
+        app_name    = 'main'
+        config_vars = {}
+        config_uri  = 'production.ini'
+
+        loader = get_config_loader(config_uri)
+        loader.setup_logging(config_vars)
+        app = loader.get_wsgi_app(app_name, config_vars)
```

Comment:
uWSGI has options `--paste` and `--ini-paste-logged` which makes `wsgi.py` redundant. You can just use this options and it will boot pyramid app ootb.

Reply:
I believe those options requires `PasteScript` dependency to run correctly though. Might be worth mentioning.

Reply:
Actually I did a little digging, and it seems `PasteScript` is only required if you like to use --ini-paste-logged and have logs forwarded from the app. Otherwise it will work without additional dependency.

Reply:
@pagenoare

It would be great to show an example that loads all the uWSGI bits
from the .ini file. As soon as I have a working version I would like to
add it to this tutorial. Here are the roadblocks I've run into so far. 


### uwsgi does not recognize --paste option

Running on my Ubuntu 16.04 box running cpython 3.6 give an error:

    $ uwsgi --paste config:/home/ubuntu/stethoscope/production.ini --socket :3031 -H /home/ubuntu/env
    uwsgi: unrecognized option '--paste'
    
    

## Where is --paste defined?

More research reveals that the "paste" option is specified in this
file in the uWSGI source code: 
https://github.com/unbit/uwsgi/blob/master/plugins/python/python_plugin.c

But is not present in core:
https://github.com/unbit/uwsgi/blob/master/core/uwsgi.c


### Invoking uWSGI with python3 plugin 

Invoking uWSGI with python3 plugin so --paste option will be recognized

    $ uwsgi --plugin python3 --paste config:/home/ubuntu/stethoscope/production.ini --socket :3031 -H /home/ubuntu/env

Gives this output:

    *** Starting uWSGI 2.0.15-debian (64bit) on [Tue Jul 10 08:04:33 2018] ***
    compiled with version: 7.3.0 on 01 March 2018 10:02:30
    os: Linux-4.15.0-1010-aws #10-Ubuntu SMP Thu May 24 08:41:34 UTC 2018
    nodename: ip-172-31-6-94
    machine: x86_64
    clock source: unix
    pcre jit disabled
    detected number of CPU cores: 1
    current working directory: /home/ubuntu/stethoscope
    detected binary path: /usr/bin/uwsgi-core
    dropping root privileges as early as possible
    *** WARNING: you are running uWSGI without its master process manager ***
    your processes number limit is 1825
    your memory page size is 4096 bytes
    detected max file descriptor number: 1024
    lock engine: pthread robust mutexes
    thunder lock: disabled (you can enable it with --thunder-lock)
    uwsgi socket 0 bound to TCP address :3031 fd 3
    dropping root privileges after socket binding
    Python version: 3.6.5 (default, Apr  1 2018, 05:46:30)  [GCC 7.3.0]
    Set PythonHome to /home/ubuntu/env
    Fatal Python error: Py_Initialize: Unable to get the locale encoding
    ModuleNotFoundError: No module named 'encodings'


### More succinctly: with --ini

I also commonly see uWSGI invoked with --ini as follows:

    $ uwsgi --ini production.ini

But so far this gives me the same "ModuleNotFoundError: No module named 'encodings'"
with this approach.

Reply:
@jackdesert 

Can you tell me which uwsgi version are you testing this on? Also, would be helpful for me, if you could provide the .ini `[uwsgi]` section. Today's morning I was testing recipe from https://docs.pylonsproject.org/projects/pyramid-cookbook/en/latest/deployment/uwsgi_nginx_systemd.html and it was working just fine.
```
$ pip freeze | grep uWSGI
uWSGI==2.0.17.1
```
$ python --version
Python 3.6.5```

OS version:
```Description:    Ubuntu 18.04 LTS```

I'm gonna try it on ubu 16.04 later today (or tomorrow morning) and see what happens.

Reply:
@jackdesert Interesting. Indeed `--paste` (or `--ini`) option doesn't work well. However, `--ini-paste-logged` one works as expected to me. I'm gonna do some more digging and let you know.

Reply:
@pagenoare 

I'm running on a fresh amazon box, the AMI is ami-39c28c41 in us-west-2, it's Ubuntu 18.04.

### uWSGI Version

    $ sudo uwsgi  --version
    2.0.15-debian

### .ini File

    [uwsgi]
    chdir = /home/ubuntu/myproject
    wsgi-file = wsgi.py
    callable = app
    processes = 2
    threads = 2
    offload-threads = 2
    stats =  127.0.0.1:9191
    max-requests = 5000
    master = True
    vacuum = True
    enable-threads = true
    harakiri = 60
    logto = /var/log/uwsgi/%n.log
    chmod-socket = 020
    plugin = python3
    pidfile=%(chdir)/tmp/%n.pid
    socket = %(chdir)/tmp/%n.sock
    virtualenv = %(chdir)/env/bin
    uid = ubuntu
    gid = www-data

### Entry Point

Using the same entry point from this pull request. 

    # Adapted from PServeCommand.run in site-packages/pyramid/scripts/pserve.py
    from pyramid.scripts.common import get_config_loader
    app_name    = 'main'
    config_vars = {}
    config_uri  = 'production.ini'

    loader = get_config_loader(config_uri)
    loader.setup_logging(config_vars)
    app = loader.get_wsgi_app(app_name, config_vars)


### Invoking uWSGI

    $ sudo uwsgi --ini production.ini 
    [uWSGI] getting INI configuration from production.ini
    Aborted

### Log Messages

Here are the last four log messages:

    Python version: 3.6.5 (default, Apr  1 2018, 05:46:30)  [GCC 7.3.0]
    Set PythonHome to /home/ubuntu/myproject/env/bin
    Fatal Python error: Py_Initialize: Unable to get the locale encoding
    ModuleNotFoundError: No module named 'encodings'

Reply:
@pagenoare 

### Good News

Fixed the "No module named 'encodings'" error by changing the "virtualenv" setting in the .ini file from 

    %(chdir)/env/bin 

to 

    %(chdir)/env. 

### New .ini File

Here is the updated [uwsgi] section of the .ini file. The `virtualenv` is the big needed change, 
but also changed from using %n to using %c.

    [uwsgi]
    chdir = /home/ubuntu/%c
    wsgi-file = wsgi.py
    callable = app
    processes = 2
    threads = 2
    offload-threads = 2
    stats =  127.0.0.1:9191
    max-requests = 5000
    master = True
    vacuum = True
    enable-threads = true
    harakiri = 60
    logto = /var/log/uwsgi/%c.log
    chmod-socket = 020
    plugin = python3
    pidfile=%(chdir)/tmp/%c.pid
    socket = %(chdir)/tmp/%c.sock
    virtualenv = %(chdir)/env
    uid = ubuntu 
    gid = www-data

### Renders with --ini

Invoking using sudo and --ini

    $ sudo uwsgi --ini production.ini

Shows good news in the log file:

        WSGI app 0 (mountpoint='') ready in 1 seconds on interpreter 0x55787475eff0 pid: 3477 (default app)

And works with Nginx

    $ curl localhost
    ... (bunch of html)

Reply:
@stevepiercy @pagenoare 
### Another Section

Now that that's working I'm looking forward to adding another section to this
tutorial that demonstrates invoking the pyramid app as

    sudo uwsgi --ini production.ini

### On Whether a Tutorial Should Show One Way or Both Ways of Invoking

Of the two methods (let's call one "--ini" and the other "--many-args"), the "--ini" version seems easier to maintain, and I expect folks will gravitate to using it. The "--ini" version also aligns with deploying from emperor, which is my (personal) goal.

Yet the "--many-args" version seems more explicit, less magical, and its inclusion makes the tutorial more complete. That's why I'm leaning toward showing both methods. If you disagree, let me know.

@stevepiercy I'd be happy to see the commits thus far be shipped out the door as a unit of work. That is, unless you have a strong inkling to wait till the next installment is ready.

Reply:
There's no rush. I think it would be better to announce it when it is complete.

When you show both methods, suggest adding an introduction section heading (one level below chapter/primary heading) that discusses use cases for each method and why one may be more appropriate for certain situations. Then document each method, each under their own section subheading (two levels below).

Reply:
@jackdesert 

Nice findings. To me, personally, maintaining only .ini file is way better/easier than keeping wsgi.py file in the project. But for sure - its nice to have an example of such file. I would, however, post this example as optional way and point people into using .ini configuration.